src/mpi_cuda/gather.hpp

```diff
@@ -0,0 +1,162 @@
+class GatherAlState : public AlState {
+public:
+  GatherAlState(const T* sendbuf, T* recvbuf, size_t count, int root,
+                MPICUDACommunicator& comm, cudaStream_t stream) :
+    AlState(nullptr),
+    rank_(comm.rank()), root_(root), count_(count),
+    host_mem_(get_pinned_memory<T>(rank_ == root_ ? comm.size()*count_ : count_)),
+    d2h_event_(cuda::get_cuda_event()),
+    h2d_event_(rank_ == root_ ? cuda::get_cuda_event() : nullptr),
+    comm_(comm.get_comm()) {
+
+    bool const i_am_root = rank_ == root_;
+    bool const inplace_operation = sendbuf == recvbuf;
+
+    // Transfer data from device to host and use an event to determine when it
+    // completes.
+    if (i_am_root)
+      AL_CHECK_CUDA(cudaMemcpyAsync(
+                      host_mem_+rank_*count_,
+                      inplace_operation ? sendbuf+rank_*count_ : sendbuf,
+                      sizeof(T)*count, cudaMemcpyDeviceToHost, stream));
+    else
+      AL_CHECK_CUDA(cudaMemcpyAsync(host_mem_, sendbuf, sizeof(T)*count_,
+                                    cudaMemcpyDeviceToHost, stream));
+    AL_CHECK_CUDA(cudaEventRecord(d2h_event_, stream));
+
+    // Enqueue the kernel to wait on the host; root only
+    if (i_am_root) {
+      gpuwait_.wait(stream);
+
+      // Transfer completed buffer back to device.
+      AL_CHECK_CUDA(cudaMemcpyAsync(recvbuf, host_mem_, sizeof(T)*count*comm.size(),
+                                    cudaMemcpyHostToDevice, stream));
+      AL_CHECK_CUDA(cudaEventRecord(h2d_event_, stream));
+    }
+  }
+
+  ~GatherAlState() override {
+    release_pinned_memory(host_mem_);
+    cuda::release_cuda_event(d2h_event_);
+    if (h2d_event_) {
+      cuda::release_cuda_event(h2d_event_);
+    }
+  }
+
+  bool step() override {
+    if (!gather_started_) {
+      // Check if mem xfer complete
+      cudaError_t r = cudaEventQuery(d2h_event_);
+      if (r == cudaSuccess) {
+        if (root_ == rank_)
+          MPI_Igather(MPI_IN_PLACE, count_, mpi::TypeMap<T>(),
+                      host_mem_, count_, mpi::TypeMap<T>(),
+                      root_, comm_, &req_);
+        else
+          MPI_Igather(host_mem_, count_, mpi::TypeMap<T>(),
+                      host_mem_, count_, mpi::TypeMap<T>(),
+                      root_, comm_, &req_);
+        gather_started_ = true;
+      }
+      else if (r == cudaErrorNotReady) {
+        return false;
+      }
+      else {
+        throw_al_exception("Alltoall: cudaEventQuery error");
+      }
+    }
+
+    if (!gather_done_) {
+      // Wait for the all2all to complete
+      int flag;
+      MPI_Test(&req_, &flag, MPI_STATUS_IGNORE);
+      if (flag) {
+        gather_done_ = true;
+        if (rank_ == root_)
+          gpuwait_.signal();
+        else
+          return true;
+      }
+      else {
+        return false;
+      }
+    }
+    else if (rank_ != root_) {
+      // Paranoia, in case step() is ever called again after returning
+      // 'true' for the first time.
```

Comment:
If that happens, I'd call it a bug in the progress engine, but fair enough.

Reply:
I agree, but it's a simple enough check...